### PR TITLE
Add tests for postsWithLighthouse collection

### PR DESCRIPTION
**Pull request by devnook**

### CODE REVIEW

**Comment by robdodson** [comment]
Instead of running eleventy I think it might be better to mock the `collection` object [that eleventy passes to the collection function](https://github.com/GoogleChrome/web.dev/blob/master/src/site/_collections/posts-with-lighthouse.js#L22).

File: test/site/_collections/posts-with-lighthouse/pages/collection.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{%- for post in collections.postsWithLighthouse -%}
```
Comment (robdodson) [inline test/site/_collections/posts-with-lighthouse/pages/collection.md:1]: What do you think about renaming this directory to `fixtures` so folks can tell at a glance what these are

File: test/site/_collections/posts-with-lighthouse/dev.spec.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  '..',
+  '.tmp',
+  path.basename(__dirname),
+  'pages',
+  'collection',
+  'index.html',
+);
+
+describe('posts-with-lighthouse', function() {
+  describe('postsWithLighthouse', function() {
```
Comment (robdodson) [inline test/site/_collections/posts-with-lighthouse/dev.spec.js:16]: not sure if you need this second level of nesting, you could probably do:

```
describe('posts-with-lighthouse', function() {
  it('...')
});
```

File: test/site/_collections/posts-with-lighthouse/dev.spec.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    it('creates postsWithLighthouse collection in dev env', async function() {
```
Comment (robdodson) [inline test/site/_collections/posts-with-lighthouse/dev.spec.js:17]: I think the message may be redundant because mocha will be showing:

```
_collections
  in DEV env
    posts-with-lighthouse
      creates postsWithLighthouse collection in dev env
```

One question to ask: what aspect of this output do we care about? Did it include some files and exclude others? Maybe we would say that it output the expected number of files with the `pathItem` tag and the `web_lighthouse` field?

You could also add test fixtures that have `web_lighthouse: N/A` and maybe junk stuff like `web_lighthouse: true` or null, etc.

File: test/site/_collections/posts-with-lighthouse/prod.spec.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  '..',
+  '.tmp',
+  path.basename(__dirname),
+  'pages',
+  'collection',
+  'index.html',
+);
+
+describe('posts-with-lighthouse', function() {
+  describe('postsWithLighthouse', function() {
```
Comment (robdodson) [inline test/site/_collections/posts-with-lighthouse/prod.spec.js:16]: See above about nesting these `describe()` calls.

File: test/site/_collections/posts-with-lighthouse/prod.spec.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    it('does not include drafts in the postsWithLighthouse collection in prod', async function() {
```
Comment (robdodson) [inline test/site/_collections/posts-with-lighthouse/prod.spec.js:17]:
```suggestion
    it('does not include drafts', async function() {
```

File: test/site/_collections/index.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    });
+
+    after(function() {
+      fs.rmdirSync(path.join('.', '.tmp'), {recursive: true});
+    });
+
+    require('./posts-with-lighthouse/prod.spec');
+    require('./recent-posts/prod.spec');
+  });
+});
```
Comment (robdodson) [inline test/site/_collections/index.js:33]: Note there's also a `PERCY` env that changes the behavior of `livePosts`. We should test that as well.

File: test/site/_collections/recent-posts/dev.spec.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  '.tmp',
+  path.basename(__dirname),
+  'pages',
+  'collection',
+  'index.html',
+);
+
+describe('recent-posts', function() {
+  describe('recentPosts', function() {
+    it('creates recentPosts collection in dev env', async function() {
```
[the remark this answers is not in the document]
Comment (robdodson) [reply test/site/_collections/recent-posts/dev.spec.js:17]: Also might be worth highlighting that it excludes posts if they don't have a hero or thumbnail. Might also state that it limits it to the correct number of posts (3)

File: test/site/_collections/recent-posts/prod.spec.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  '..',
+  '.tmp',
+  path.basename(__dirname),
+  'pages',
+  'collection',
+  'index.html',
+);
+
+describe('recent-posts', function() {
+  describe('recentPosts', function() {
```
Comment (robdodson) [inline test/site/_collections/recent-posts/prod.spec.js:16]: see note above about nesting

File: test/site/_collections/recent-posts/prod.spec.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    it('excludes drafts in the recentPosts collection in prod', async function() {
```
Comment (robdodson) [inline test/site/_collections/recent-posts/prod.spec.js:17]:
```suggestion
    it('excludes drafts', async function() {
```

File: test/site/_collections/eleventy.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+const path = require('path');
+const postsWithLighthouse = require(`../../../src/site/_collections/posts-with-lighthouse`);
+const recentPosts = require(`../../../src/site/_collections/recent-posts`);
+
+module.exports = function(config) {
+  config.addCollection('postsWithLighthouse', postsWithLighthouse);
+  config.addCollection('recentPosts', recentPosts);
```
[the remark this answers is not in the document]
Comment (devnook) [reply test/site/_collections/eleventy.js:7]: wow!

File: test/site/_collections/index.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+const fs = require('fs');
+const path = require('path');
+const runEleventy = require('./runEleventy');
+
+describe('_collections', function() {
+  describe('_collections in DEV env', function() {
```
[the remark this answers is not in the document]
Comment (devnook) [reply test/site/_collections/index.js:6]: done

File: test/site/_collections/posts-with-lighthouse/pages/collection.md (hunk 1 above)
Comment (devnook) [reply test/site/_collections/posts-with-lighthouse/pages/collection.md:1]: done

File: test/site/_collections/index.js (hunk 6 above)
Comment (devnook) [reply test/site/_collections/index.js:33]: Noted. Will try to test in in livePosts test then.

File: test/site/_collections/runEleventy.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+const assert = require('assert');
+const util = require('util');
+const exec = util.promisify(require('child_process').exec);
+
+module.exports = async function(env) {
+  console.log('Running Eleventy...');
+  try {
+    await exec(
+      `ELEVENTY_ENV=${env} npx @11ty/eleventy --config="./eleventy.js"`,
```
[the remark this answers is not in the document]
Comment (devnook) [reply test/site/_collections/runEleventy.js:9]: npx should pick up eleventy from our node_modules dir, so should be compatible with our main run. If I run `npm run eleventy` here, it is run from the / root dir i picks up the evelenty.js config located there, even if I provide the correct path later on (which is weird). I can also run directly `../../../../../node_modules/.bin/eleventy --config="./eleventy.js" ` but slightly prefer npx as it will get dependencies in case some modules are missing. Wdyt?

**Comment by devnook** [comment]
PTAL

**Comment by paulirish** [comment]
Just curious, what does this collection do?  I'm somewhat of an 11ty noob.

**Comment by robdodson** [comment]
@paulirish it looks at all of the markdown files that have `web_lighthouse` in the YAML front matter and puts them into an array. Later on that array gets dumped into a big json that's at the bottom of the /measure/ page. When the measure tool runs it uses that json to know which guide to link to. It's a super old system from the earliest days of the site, before we had plans to just write audit guides for all of LH.

Now that we have audit guides for LH it would probably make sense to redo this but it works for the moment.

**Comment by paulirish** [comment]
@robdodson super helpful thanks!
